staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -710,19 +711,19 @@ func TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter(t *testing.T) {
 		//  - first request is expected to panic as designed
 		//  - second request is expected to success
 		_, err := requestGetter(firstRequestPathPanic)
-		if !executed {
+		if invokedGot := atomic.LoadInt32(&invoked); invokedGot != 1 {
```

Comment:
Hmm - if we had a race before [so something was writing it at the same time when something else was reading], won't this now effectively become flaky [sometimes it will be 1, sometimes something else]?

Reply:
I guess my point was slightly different - if we have long-enough time between the timeouts, then why the rate here was happening in the original version?

Either there is enough time in between (and then the original version should work), or there isn't enough time and then with the new code you could be racy too, no?

Reply:
here is my observation so far:
- the data race does not manifest with the race detector on, i have tried with `stress`, no luck so far
```
$ stress -p=32 ./filters.test -test.run=TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter
5s: 0 runs so far, 0 failures
10s: 32 runs so far, 0 failures
...
15m30s: 4482 runs so far, 0 failures
15m35s: 4505 runs so far, 0 failures
```

- with https://github.com/kubernetes/kubernetes/pull/114189, when read/write deadline is enabled, this data race manifests.

I wrote a much simpler, but equivalent test with the race present, `executed` variable being written to by the server filter goroutine, and read by the test goroutine, and if I run this test the data race is detected, but for some reason the same race is not detected in the `TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter` test.
```
func TestDataRaceWithServerHandler(t *testing.T) {
	handlerDoneCh, clientDoneCh := make(chan struct{}), make(chan struct{})
	var executed bool
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer close(handlerDoneCh)
		executed = true

		<-clientDoneCh
	})

	server := httptest.NewUnstartedServer(handler)
	defer server.Close()
	server.EnableHTTP2 = true
	server.StartTLS()

	client := server.Client()
	client.Timeout = 5 * time.Second

	func() {
		defer close(clientDoneCh)
		client.Get(server.URL + "/foo")
	}()

	if !executed {
		t.Errorf("expected handler to be executed")
	}
}
```

due to #114189, net/http will spin up a new goroutine for each request on the server side, which directly returns a stream reset error to the caller. So it brings in two new changes:
- a new goroutne in play (but it does not directly read from or write to the memory)
- it helps the client `Get` call to done  sooner (before, the response would come to the caller via the timeout filter)
in summary, the writer (test goroutine) writes to the memory location sooner with #114189, but I am not sure why this reduced gap in read-write would manifest the data race. 

This PR does two things:
- move both read and write in the same goroutine (move t.Errorf inside the server filter, as it is safe to do so now)
- use channel instead of atomic, this minimizes the chance of a flake.